Document errors 8400, 8401, 8403, 8410, and 8411 (#19441)
dotnet/docs: docs/csharp/language-reference/compiler-messages/cs8410.md

---
title: Compiler Error CS8410
ms.date: 07/11/2020
f1_keywords:
  - "CS8410"
helpviewer_keywords:
  - "CS8410"
author: Youssef1313
---
# Compiler Error CS8410

'type': type used in an asynchronous using statement must be implicitly convertible to 'System.IAsyncDisposable' or implement a suitable 'DisposeAsync' method.

The expression inside an `await using` statement must have a `DisposeAsync` method.

## To correct this error

> Use `using` instead of `await using`.

Reviewer: This is not the correct messaging for this error message. This actually corresponds to `ERR_NoConvToIAsyncDispWrongAsync = 8417`, see [resource](https://github.com/dotnet/roslyn/blob/dbdae0b756975b86f3411c3ec9e01df4afbe1fc8/src/Compilers/CSharp/Portable/CSharpResources.resx#L3004-L3009)

Author: @IEvangelist, this line isn't referred to as being the message of the error. It just suggests the common fix for this error.

Reviewer: [See this please.](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMBWAbl1xIDZi1iMB2Q3Ab10K8bpQA5idALIQAlgDsAFAEpWOboQ7yF3FAE5iRAG4QEhBHAgATRIQC8hcXADuhAEqGTCGZWUriGgoR16joqAAOAPZQEMAANnAOxqYWVrYAIv7BoRFRjogunO5yOZreuoQAHuaWNvYZztKueV4+hACepfGESYEhYZHRTlluhAC+uIM41AzdpmwDVHgMbSmd6TF6IIQAknMdaezZXCgAzMQALK3JIXAyilM4/UA=)

Author: Oh got it. Then what you recommend to put under "To correct this error"?

Reviewer: The only thing that makes sense is to remove both the `await` and the `using`. If the type is neither an implementation of `IDisposable` or `IAsyncDisposable`, I think that's the only logic suggestion that can be made. @BillWagner, do you agree?

Reviewer 2: Yes, I agree. If a type is not disposable (synchronously or asynchronously), the only fix is to remove the `using`

Author: @BillWagner @IEvangelist

I think we may also add "implement a suitable `DisposeAsync` method" as another option. What do you think?

Reviewer: Here is an updated version of the [SharpLab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMBWAbl1xIDZi1iMB2Q3Ab10K8bpQA5idALIQAlgDsAFAEpWOboQ7yF3APSrCAYQDKfACwYiAWkIwAFqKiFLhAHIB5ACqEAZqIAeEYABs4hAO6i5qZmfghwYAD2AG4Q3oSRLoQABigAnMmcKlzpxESxCIThEAAmiIQAvITicP6EAEpwpYgylMrZWdnqdpGEiAiRCJ0qBUVNZQgMVTV1jc0IrcMKS2oaOvrMhCbmNjZunj5+gcHmYRExcQlJqRkrOWl5hKMllgAOkVBevnMTldW1hAAIm8Pl84D8WtI2tkuHdCN1bL1+oM4QQnhBCi8oO9PocIZM/jMgSDcd9xpDodwAL64Gk4agMfGKQh0hnE7GgvHkwogQgASWBHNJcHYwxQAGZiHp2Ti4DJmXSqUA===)

Author: These are very useful when documenting cs8417. We may even link the two pages as they're a bit related.

## Example

> <!-- TODO: The current example is a little bit poor. -->

Reviewer 2: I think a really good example would be using `StreamReader`, because `StreamWriter` does have a `DisposeAsync` method, but `StreamReader` does not.

Author: I tried StreamReader but it generated CS8417 instead of CS8410 (Both errors are similar). Not sure when the compiler generates which one.

Reviewer: Is this supposed to be an example of code that doesn't compile? If so, you'd show an example of something that is errant.

[Here is a SharpLab.io] (https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMA6ASQHkBuXA4jAVlpzowDZi0GB2Q3AN65CIhpxQAOYpwCyEAJYA7ABQBKfjlGEhmraJQBOYkWUA3CAkII4EACaJCAXkKK4Ad0IBlGNYhgASjb2CMoARKGq6gD0UYQAIgD2cFCKAOQwhAAWEKZwhPIwwnoiMYbGhGYWVkEOzq4e3r5gAOoIBYhhEdGx2VDx8lAADglQcACCUACeigDGRcU6xSIAvvOEqzjLQA) showing that the `StreamWriter` can be used with `await using` but the `StreamReader` cannot.

Author: @IEvangelist Yes the examples are of code that doesn't compile, but it's supposed to generate a **specific** compile error. In this specific one, we need CS8410, but with StreamReader it generates CS8417. I'm not sure why the compiler team made these two different error code while they look just the same for me. It's unclear to me when the compiler would generate CS8410 and when it would generate CS8417.

Reviewer: Let me see what I can dig up

Reviewer: Hey @Youssef1313 - try this [SharpLab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGArAbgFgAoEnANiwCYtsB2DEgbxIzdspQA4tKBZCAEsAdgAoAlI2LsMLaTPYoAnFkwA3CAgwI4EACaIMAXgzC4AdwwBJMAAcANoIDGgmACVdBhBKLy2AXxJA0mIUGhsHZ1cPfURmIKA===).

Author: @IEvangelist, Your example shows the error message we want here. But any idea why StreamReader generates a different message?

Reviewer: The reason is that the `StreamReader` implements `IDisposable`, this goes back to my earlier comment that removing the `await` is not valid for `8410` as a suggestion to fix it. That only applies to 8417.

[See this](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMBWAbl1xIDZi1iMB2Q3Ab10K8bpQA5idALIQAlgDsAFAEpWOboQ7yF3FAE5iRAG4QEhBHAgATRIQC8hcXADuhAEqGTCGZWUriGgoR16joqAAOAPZQEMAANnAOxqYWVrYAIv7BoRFRjogunO5yOZreuoQAHuaWNvYZztKueV4+hACepfGESYEhYZHRTlluhAC+uIM41AzdpmwDVHgMbSmd6TF6IIQAknMdaezZXCgAzMQALK3JIXAyilM4/UA=).

```csharp
using System;
using System.Threading.Tasks;

static class Program 
{
    static async Task Main() 
    {
        await using var reader = new Reader();
        await using var disposableReader = new DisposableReader();
        
        using var x = new Reader();
        using var y = new DisposableReader();
    }
}

class Reader { }

class DisposableReader : IDisposable
{
    public void Dispose() { }
}
```

```csharp

> class C
> {
>     public static async System.Threading.Tasks.Task Main()
>     {
>         await using (Main)
>         {
>         }
>     }
> }

Reviewer: ```suggestion
using System.Threading.Tasks;

static class Program 
{
    static async Task Main() 
    {
        // error CS8410: 'ImplicitReader': type used in an asynchronous using statement
        // must be implicitly convertible to 'System.IAsyncDisposable' or implement
        // a suitable 'DisposeAsync' method.
        await using var reader = new ImplicitReader();
    }
}

class ImplicitReader
{
}
```

> ```

Reviewer: Consider adding a see also and linking to the Implement DisposeAsync article.

```markdown
## See also

- [Implement a DisposeAsync method](../../../standard/garbage-collection/implementing-disposeasync.md)
```

Author: Sure. I even opened https://github.com/dotnet/docs/issues/19440 specifically for this. But looks like I forgot 😄

---
title: Compiler Error CS8410
ms.date: 07/11/2020
f1_keywords:
  - "CS8410"
helpviewer_keywords:
  - "CS8410"
author: Youssef1313
---
# Compiler Error CS8410

'type': type used in an asynchronous using statement must be implicitly convertible to 'System.IAsyncDisposable' or implement a suitable 'DisposeAsync' method.

The expression inside an `await using` statement must have a `DisposeAsync` method.

## To correct this error

Remove the `await using` keywords, or implement a suitable `DisposeAsync` method.

## Example

```csharp
using System.Threading.Tasks;

> static class Program

Reviewer: ```suggestion
class Program
```

Author: I think it's okay to keep the static keyword? that class contains only a static main method.

Reviewer: But it is not necessary. In fact to be consistent with the templates of a new C# console application, it should be omitted.

> ```csharp
> using System.Threading.Tasks;
> 
> static class Program
> {
>     static async Task Main()
>     {
>         // error CS8410: 'Example': type used in an asynchronous using statement
>         // must be implicitly convertible to 'System.IAsyncDisposable' or implement
>         // a suitable 'DisposeAsync' method.
>         await using var reader = new Example();
>     }
> }
> 
> class Example
> {
> }
> ```

Reviewer: ```suggestion
```csharp
using System.Threading.Tasks;

static class Program
{
    static async Task Main()
    {
        // error CS8410: 'Example': type used in an asynchronous using statement
        // must be implicitly convertible to 'System.IAsyncDisposable' or implement
        // a suitable 'DisposeAsync' method.
        await using var example = new Example();
    }
}

class Example
{
}
```
```
